topics/dev/tutorials/debugging/tutorial.md

```diff
@@ -0,0 +1,730 @@
+>    >  > 
+>    >  > ```python
+>    >  >     def get(self, uri):
+>    >  >         if uri not in self._by_index:
+>    >  >             return self._by_index[uri]
+>    >  >             ...
+>    >  > ```
+>    >  > Let's rewrite the first clause in pseudo code to expose the problem:
+>    >  > ```
+>    >  > if key not in dictionary: 
+>    >  >     return dictionary[key]
+>    >  > ```
+>    >  > Well, that makes no sense whatsoever! It should be the opposite: if key **is** in dictionary, then get the item with that key. The `not` is clearly a typo (or, more accurately, a wicked edit made for the purpose of setting up a learning experience). Remove this negation and rerun the test.  
+>    >  > 
+>    > {: .solution }
+>    > 
+> 4. __Finishing up__
+>    > 
+>    > Well, that was fun. Congratulations! Now commit your edit (don't forget to supply a useful commit message), and move on to client errors.
+{: .hands_on}
+
+# Client linting error
+
+Client Linting is important to reduce errors, improve quality of the codebase, and enforce common rules. These rules are defined by the eslinter defined with Vue best practices, primarily. On Github, the command `eslint` is run to start the process for linting code on the client side.
+
+> ### {% icon comment %} Python Linting
+> For backend consistency, we also have a Python linter. While we won't be running you through that exercise today, you may encounter a Python linting issue in the future. The way to diagnose and treat these is very similar to the way to do it for the front end.
+{: .comment}
+
+> ### {% icon hands_on %} Hands-on: Fixing client linting error
+> 
+> 1. __Finding the failing test on GitHub__
+> 
+>    > One of the failing tests on GitHub says "get_code_and_test". Clicking on the word "Details" assocated with this failure will open a brief view of the CircleCI Checks, with a list of tests that are run as part of that command. You'll see that the one that is failing is js_lint. And if you click that, you'll be navigated to the CicleCI app, where you can see much more information about the failures.
+>    > 
+>    > Here you can see the exact command that ran to lint our codebase; it's `cd client && yarn run eslint`.
+>    > 
+>    > We can also see that it has found 1 error, where that error is located, and a description of the problem.
+>    > 
+> 2. __Running the linter locally__
+>    > 
+>    > Alright, so we've seen the failure, but we want to verify it locally now. For this, you'll need to open up a terminal.
+>    > 
+>    > Make sure your virtual environment is activated. If not, activate it: `. .venv/bin/activate`
+>    > 
+>    > Then you've got two options. You can run the same exact command that CircleCI uses `cd client && yarn run eslint` or `make client-lint`, which does `cd client && yarn run eslint && yarn run prettier-check`. This will also make sure your code conforms to formatting standards.
```

Comment:
Hmm, I always took it for granted: I thought it was installed automagically! If not then yes, sure, we should include that!

Reply:
I'm pretty sure (though, by no means an expert & have great potential to be wrong) that yarn is installed as part of the virtual environment, which would make it unnecessary to install it first.

Someone -- fact check me.

Reply:
You are right, the documentation says:

> The base dependencies used are Node.js and Yarn. Galaxy now includes these in
the virtual environment, and they can be accessed by activating that with `.
.venv/bin/activate` from the Galaxy root directory.

I don't know what went wrong then, I checked the tutorial in a freshly installed Ubuntu and I would have sworn that the virtual environment was activated (as I carefully followed every step) but, in my case, I have **huge** potential to be wrong :laughing:  so this can be safely ignored, as I said, is not a big deal and `make client-lint` points you to the installation instructions in case someone steps into it :)

Reply:
Thank you for checking and verifying!